### ISSUE TITLE
Unable to install Kubeapps in single-cluster mode with OIDC

### ISSUE THREAD

**Issue by antgamdia** [kubeapps/kubeapps#2692]
Follow-up of #2662

### Description:

Unable to install Kubeapps in single-cluster mode with OIDC, for instance, when installing Kubeapps with a non-native OIDC provider in managed clusters (e.g., using VMware CSP login in an AKS cluster).

### Steps to reproduce the issue:

_TBD: I want to test if this issue also applies when not using Pinniped. I'll update this issue in a while._

1. Get access to a managed k8s cluster (we use AKS henceforth)
2. Install Pinniped 0.7 in AKS
3. Retrieve the impersonation proxy URL and CA following our guides
4. Install Kubeapps using a similar config:
```yaml
pinnipedProxy:
  enabled: true

authProxy:
  additionalFlags:
  - --oidc-issuer-url= #your oidc provider
  - --scope=openid email groups
  clientID: <redacted>
  clientSecret: <redacted>
  cookieSecret: <redacted>
  enabled: true
  provider: oidc

clusters:
- apiServiceURL: https://<IMPERSONATION_PROXY_URL>
  certificateAuthorityData: <IMPERSONATION_PROXY_CA>
  name: impersonation-proxy
  pinnipedConfig:
    enable: true
```
3. Install kubeapps as usual
3. Log in and perform actions.

#### Not able to deploy:

![image](https://user-images.githubusercontent.com/11535726/115266968-f5dbdf00-a138-11eb-9955-14c5f3a43170.png)

Using URL:

http://IP/api/kubeops/v1/clusters/impersonation-proxy/namespaces/default/releases

#### Not able to rollback:
![image](https://user-images.githubusercontent.com/11535726/115279647-435f4880-a147-11eb-939f-2524442ca763.png)
> This is odd, seems an UI issue?
> 1. the URL is http://IP/api/kubeops/v1/clusters/impersonation-proxy/namespaces/default/releases/nginx-test-to-aks?action=rollback&revision=-1
> 2. The it becomes http://IP/api/kubeops/v1/clusters/impersonation-proxy/namespaces/default/releases/nginx-test-to-aks?action=rollback&revision=NaN

#### Not able to list repositories

Using URL:

http://IP/api/v1/clusters/impersonation-proxy/namespaces

#### Able to upgrade
Using URL: 

http://IP/api/kubeops/v1/clusters/impersonation-proxy/namespaces/default/releases/nginx-test-to-aks

#### Able to create and list namespaces

Using these URLs:

http://IP/api/v1/clusters/impersonation-proxy/can-i
http://IP/api/v1/clusters/impersonation-proxy/namespaces

### Describe the results you expected:

Kubeapps should use the proper endpoints instead of relying on the `inCluster` k8s API configuration.

### Additional information you deem important (e.g. issue happens only occasionally):

Gut feeling: it is related to the usage of `pinnipedConfig.enable=true`

While writing the steps, it seems there is an issue with some URLs (e.g., can-i endpoint). We should address these first and later see if there is also a backend issue.

As per https://github.com/kubeapps/kubeapps/issues/2662#issuecomment-822349783:

> So this is interesting: when I demo'd Kubeapps running in a TKG (1.1) management cluster I had this exact scenario, if I've understood correctly: Kubeapps was installed on the management cluster but was configured to deploy to the workload cluster(s) only (because TKG 1.1 didn't support OIDC for the management cluster, only the workload clusters). That is, it was configured with (both) clusters with explicit api urls and certs. So I'm keen to understand what changed that broke this (though I guess it's not surprising since we hadn't added CI tests). Note that the inCluster config is (from memory) only used to initialise the variable, I thought we overwrite the endpoint based on the config.

### Version of Helm, Kubeapps and Kubernetes:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.5.2", GitCommit:"167aac70832d3a384f65f9745335e9fb40169dc2", GitTreeState:"dirty", GoVersion:"go1.15.7"}
```

- Output of `helm list -n <kubeapps-namespace> <kubeapps-release-name>`:

```
NAME            NAMESPACE       REVISION        UPDATED                                         STATUS          CHART                APP VERSION
kubeapps        kubeapps        2               2021-04-16 15:28:36.759785314 +0200 CEST        deployed        kubeapps-6.0.1       DEVEL      
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-11T13:17:17Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3+vmware.1", GitCommit:"2ac9e7ea06a1230ca196931def19d2bb67b580c7", GitTreeState:"clean", BuildDate:"2020-10-30T07:25:38Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by antgamdia** [comment]
Well, one issue seems to be here:

https://github.com/kubeapps/kubeapps/blob/fefc870a1d39e09fab5a93c0c2c69b7baf1a64e9/cmd/kubeops/main.go#L204-L210

In those features exclusives to "the cluster in which Kubeapps is installed" we are relying on `configs.KubeappsClusterName`. The issue is that we are supposing we always have a cluster without `APIServiceURL`. I'm thinking of two alternatives:
1. Just add a fallback mechanism (eg. use the first cluster). Drawback: prevent users from knowing what's happening (ie, which cluster has been chosen) + issues when deserializing yaml (is the order guaranteed?, we should check it)
2. add a field: `default: true` so that users can decide which cluster they want to use.
3. Combine both: if `default`, use it; if not, look for the one without API URL, if not, use the first one.

WDYT @absoludity ?

**Comment by absoludity** [comment]
Let's limit the discussion to the error in the screenshot: `unable to create clientset: cluster "" has no configuration`

> Well, one issue seems to be here:
> 
> https://github.com/kubeapps/kubeapps/blob/fefc870a1d39e09fab5a93c0c2c69b7baf1a64e9/cmd/kubeops/main.go#L204-L210
> 
> In those features exclusives to "the cluster in which Kubeapps is installed" we are relying on `configs.KubeappsClusterName`. The issue is in supposing we always will have a cluster without `APIServiceURL`.

I think the direct cause of the raised error is that when creating a release, `.AsSvc()` is being called (via `chart.GetAppRepoAndRelatedSecrets()`)in the kubeops code with an empty string for the cluster. The `.AsSvc()` helper that returns an http client for accessing the k8s API of the cluster on which kubeapps is itself running, was updated 6 months ago to require an optional `cluster` arg so it could be used to fetch operator icons on other clusters via a service account, with the existing call-sites updated to use the `kubeappsClusterName` as the argument. See the [change in 2091](https://github.com/kubeapps/kubeapps/pull/2091/files#diff-680a1eb150fd209972a67711ba74bb11159483a38c054070ca2d01d8dc919007R180). Prior to this, `.AsSvc()` assumed that since kubeops *is* part of the Kubeapps installation on that cluster, it can use the in cluster config. As such, there was never any expectation that `kubeappsClusterName` would be anything but empty if that cluster wasn't an install target for the user (ie. if the user wasn't able to select that cluster in the UI). The names defined in the cluster configs are after all, just user-facing labels for target clusters and the `kubeappsClusterName` is (was) only relevant if the kubeapps cluster was itself one of the targets. This is the reason why the TKG install on a management cluster targeting workload clusters only, used to work but perhaps would not now, since the empty label is now passed and expected to be part of the configuration. It is also the cause of the actual error being raised in this case, since `kubeappsClusterName` is empty for a different reason, but nonetheless, I think updating the [`.AsSvc()` implementation](https://github.com/kubeapps/kubeapps/pull/2091/files#diff-680a1eb150fd209972a67711ba74bb11159483a38c054070ca2d01d8dc919007R235) to revert to the old behavior if the kubeapps cluster name is empty is worthwhile (and would avoid the noted error when deploying).

That said, in the case you've got here, we *do* have a cluster in the configuration which is the one on which Kubeapps is installed, since it is also the one users should target for deploys, but it is not being recognised as the kubeapps cluster as you've mentioned. This is because up until now, the K8s API server URL was always the internal one (`https;//kubernetes.default`). A question:

1. If we fix this so that `kubeappsClusterName` is correctly set in the config even with the url for the impersonation proxy, is the impersonation proxies LB actually what we want in this specific case of `.AsSVC()`?. I think we'd still want to use `https://kubernetes.default` for the service account based access, not the impersonation proxy LB? That is, we may need to update the suggested fix for `.AsSVC()` above to be something like `if cluster == "" || cluster == kubeappsClusterName: use the in cluster config url (which will be  https://kubernetes.default`)

Regarding the fix to ensure we *can* identify the cluster on which kubeapps is installed (assuming we need to for other UI/UX conditions)...

> I'm thinking of two alternatives:
> 
>     1. Just add a fallback mechanism (eg. use the first cluster). Drawback: prevent users from knowing what's happening (ie, which cluster has been chosen) + issues when deserializing yaml (is the order guaranteed?, we should check it)

I think this option ignores the case where the clusters config does not include the cluster on which Kubeapps was deployed (remember, the config has info about the clusters a user can target, which may not necessarily include the cluster on which Kubeapps was installed).

> 
>     2. add a field: `default: true` so that users can decide which cluster they want to use.

I'm not sure why `default` would be a good key for this? We need a way to identify that this cluster which the user can target is also the cluster on which Kubeapps is installed, without relying on the absence of an api url. Without any extra data there is no way to differentiate between a single cluster setup targeting an external cluster and a single cluster with kubeapps but needing the pinniped impersonation proxy. I can't think of a better option than a new field, but perhaps a name like `isKubeappsCluster` which would only currently be needed in this case, as we'd still default to the one without an API URL.

> 
>     3. Combine both: if `default`, use it; if not, look for the one without API URL, if not, use the first one.  (I'd choose this one)

Yep, but without the last segment ("if not, use the first one"). 

> 
> 
> WDYT @absoludity ?
> 
> Edit: the other issue (frontend) is here. Same underlying issue, but different place:
> 
> https://github.com/kubeapps/kubeapps/blob/fefc870a1d39e09fab5a93c0c2c69b7baf1a64e9/chart/kubeapps/templates/_helpers.tpl#L180-L195

Yep, it worked as a default as long as we could assume that `https://kubernetes.default` was the valid endpoint for the cluster on which kubeapps is installed. We'd do the same combined logic here right?

**Comment by antgamdia** [comment]
> I think the direct cause of the raised error is that when creating a release, `.AsSvc()` is being called (via `chart.GetAppRepoAndRelatedSecrets()`)in the kubeops code with an empty string for the cluster.  (...) I think updating the [`.AsSvc()` implementation](https://github.com/kubeapps/kubeapps/pull/2091/files#diff-680a1eb150fd209972a67711ba74bb11159483a38c054070ca2d01d8dc919007R235) to revert to the old behavior if the kubeapps cluster name is empty is worthwhile (and would avoid the noted error when deploying).

The PR https://github.com/kubeapps/kubeapps/pull/2696 is solving the problems I already identified in this issue. But, at the same time, what you're saying here does make sense. Let's dig into:

So `.AsSVC` is used in these places:

- `GetOperatorLogo` (pkg/http-handler/http-handler.go), which uses `cfg.Options.ClustersConfig.KubeappsClusterName`
- `GetAppRepoAndRelatedSecrets` (pkg/chart/chart.go) which takes `mux.Vars(req)["cluster"]`

Consequently, in both cases, I think that just identifying the "kubeapps cluster" will make it. In the first scenario, this value (once retrieved) is stored in `Options.ClustersConfig`. In the second, the cluster name is retrieved from the requested URL, which is obtained in the fronted by means of the `config.json` file (which is generated by Helm, so, again, the PR fixes the issue).

> 1. If we fix this so that `kubeappsClusterName` is correctly set in the config even with the url for the impersonation proxy, is the impersonation proxies LB actually what we want in this specific case of `.AsSVC()`?. I think we'd still want to use `https://kubernetes.default` for the service account based access, not the impersonation proxy LB? That is, we may need to update the suggested fix for `.AsSVC()` above to be something like `if cluster == "" || cluster == kubeappsClusterName: use the in cluster config url (which will be  https://kubernetes.default`)

Unfortunately, I'm not following you right now. Which scenario do you think of?  I've successfully tested (deploy, upgrade, delete, add repo, update repo, charts, operators) in:

1. Single AKS cluster with Kubeapps
2. Normal cluster (TMC) adding AKS as additional cluster
3. AKS with Kubeapps adding a normal cluster (TMC).

> I'm not sure why `default` would be a good key for this? 

It was late I wasn't able to come up with a better name :P

>  can't think of a better option than a new field, but perhaps a name like `isKubeappsCluster` which would only currently be needed in this case, as we'd still default to the one without an API URL. (...) Yep, but without the last segment ("if not, use the first one"). (...) We'd do the same combined logic here right?

Great, it is what I implemented in the PR in both the Kubeops comp. and the chart.

**Comment by absoludity** [comment]
> > I think the direct cause of the raised error is that when creating a release, `.AsSvc()` is being called (via `chart.GetAppRepoAndRelatedSecrets()`)in the kubeops code with an empty string for the cluster.  (...) I think updating the [`.AsSvc()` implementation](https://github.com/kubeapps/kubeapps/pull/2091/files#diff-680a1eb150fd209972a67711ba74bb11159483a38c054070ca2d01d8dc919007R235) to revert to the old behavior if the kubeapps cluster name is empty is worthwhile (and would avoid the noted error when deploying).
> 
> The PR #2696 is solving the problems I already identified in this issue.

It is solving the problem of ensuring that `KubeappsClusterName` is set when it should be set, but it also (currently, easily fixable) further invalidates a valid use case which was broken in 2091, afaict.

> But, at the same time, what you're saying here does make sense. Let's dig into:
> 
> So `.AsSVC` is used in these places:
> 
>     * `GetOperatorLogo` (pkg/http-handler/http-handler.go), which uses `cfg.Options.ClustersConfig.KubeappsClusterName`
> 
>     * `GetAppRepoAndRelatedSecrets` (pkg/chart/chart.go) which takes `mux.Vars(req)["cluster"]`

Nope, at least, afaict, `GetAppRepoAndRelatedSecrets` is called with `cfg.Options.ClustersConfig.KubeappsClusterName` (it was initially called with `mux.Vars(req)["cluster"]` once it was required for `.AsSVC()` but that caused a different error as repos aren't supported on other clusters). You can see this at:

https://github.com/kubeapps/kubeapps/blob/v2.3.1/cmd/kubeops/internal/handler/handler.go#L181

> 
> 
> Consequently, in both cases, I think that just identifying the "kubeapps cluster" will make it. In the first scenario, this value (once retrieved) is stored in `Options.ClustersConfig`. In the second, the cluster name is retrieved from the requested URL, which is obtained in the fronted by means of the `config.json` file (which is generated by Helm, so, again, the PR fixes the issue).

Yep, I'm not claiming that your PR doesn't fix the issue of the `KubeappsClusterName` config option being empty, rather, I'm saying there is a valid case where it *can* be empty, that we used to handle, but no longer handle.

> 
> > 1. If we fix this so that `kubeappsClusterName` is correctly set in the config even with the url for the impersonation proxy, is the impersonation proxies LB actually what we want in this specific case of `.AsSVC()`?. I think we'd still want to use `https://kubernetes.default` for the service account based access, not the impersonation proxy LB? That is, we may need to update the suggested fix for `.AsSVC()` above to be something like `if cluster == "" || cluster == kubeappsClusterName: use the in cluster config url (which will be  https://kubernetes.default`)
> 
> Unfortunately, I'm not following you right now.

All I meant was that the impersonation proxy URL needs to be used as the api server URL when a *user* is accessing the cluster, for sure. I wasn't sure whether the impersonation proxy URL would work validly when accessing the cluster as a service account (as we do with `.AsSVC()`) and so whether we'd need to use the normal, non-impersonation-proxied url, but based on your tests below, it looks like it's fine to use the impersonation proxy URL with a service account credentials. Great.

> Which scenario do you think of? I've successfully tested (deploy, upgrade, delete, add repo, update repo, charts, operators) in:
> 
>     1. Single AKS cluster with Kubeapps
> 
>     2. Normal cluster (TMC) adding AKS as additional cluster
> 
>     3. AKS with Kubeapps adding a normal cluster (TMC).

Great. So the extra case that I think you're not testing is a variation of your 2 and 3, where:

4. Multiple clusters (kind is fine) where Kubeapps is running on one cluster, configured to target *only* the second cluster (similar to installing Kubeapps on the TKG management cluster but only allowing users to deploy apps on the workload clusters).

In this scenario, the clusters config does not include the cluster on which Kubeapps is installed at all, yet `.AsSVC()` needs to be able to use `https://kubernetes.default` with the service account (ie. same cluster as kubeops is running on) to query the k8s API. The changes you've made make it impossible to configure in this way, and the changes in 2091 made it impossible for `.AsSVC()` to be called in this scenario without the error you saw `unable to create clientset: cluster "" has no configuration`. I think the fix will be just a small modification to your PR (removing the requirement that kubeappsClusterName is always resolved), and a separate PR that adjusts `.AsSVC()` so that if it is called with "" (ie. kubeappsClusterName isn't set), then it uses the default `https://kubernetes.default`. Anyway, let's talk through this once you've had a chance to read and fill in any details either of us are missing :)

**Comment by antgamdia** [comment]
Let me reopen this issue, since the underlying problem is not solved until we land #2712
